packages/atclient/src/metadata.c

```diff
@@ -677,263 +681,377 @@ exit: {
+  return 5 // :ttr:
+         + long_strlen(metadata->ttr);
+}
+
+size_t atclient_atkey_metadata_ccd_strlen(const atclient_atkey_metadata *metadata) {
+  if (metadata->ccd) {
+    return 9; // :ccd:true
+  }
+  return 0;
+}
+
+size_t atclient_atkey_metadata_isbinary_strlen(const atclient_atkey_metadata *metadata) {
+  if (metadata->isbinary) {
+    return 14; // :isBinary:true
+  }
+  return 0;
+}
+
+size_t atclient_atkey_metadata_isencrypted_strlen(const atclient_atkey_metadata *metadata) {
+  if (metadata->isencrypted) {
+    return 17; // :isEncrypted:true
+  }
+  return 0;
+}
+
+size_t atclient_atkey_metadata_datasignature_strlen(const atclient_atkey_metadata *metadata) {
+  return 15 // :dataSignature:
+         + metadata->datasignature.olen;
+}
+
+size_t atclient_atkey_metadata_sharedkeystatus_strlen(const atclient_atkey_metadata *metadata) {
+  return 17 // :sharedKeyStatus:
+         + metadata->sharedkeystatus.olen;
+}
+
+size_t atclient_atkey_metadata_sharedkeyenc_strlen(const atclient_atkey_metadata *metadata) {
+  return 14 // :sharedKeyEnc:
+         + metadata->sharedkeyenc.olen;
+}
+
+size_t atclient_atkey_metadata_pubkeyhash_strlen(const atclient_atkey_metadata *metadata) {
+  return 6 // :hash:
+         + metadata->pubkeyhash.olen;
+}
+
+size_t atclient_atkey_metadata_pubkeyalgo_strlen(const atclient_atkey_metadata *metadata) {
+  return 6 // :algo:
+         + metadata->pubkeyalgo.olen;
+}
+
+size_t atclient_atkey_metadata_encoding_strlen(const atclient_atkey_metadata *metadata) {
+  return 10 // :encoding:
+         + metadata->encoding.olen;
+}
+
+size_t atclient_atkey_metadata_enckeyname_strlen(const atclient_atkey_metadata *metadata) {
+  return 12 // :encKeyName:
+         + metadata->enckeyname.olen;
+}
+
+size_t atclient_atkey_metadata_encalgo_strlen(const atclient_atkey_metadata *metadata) {
+  return 9 // :encAlgo:
+         + metadata->encalgo.olen;
+}
+
+size_t atclient_atkey_metadata_ivnonce_strlen(const atclient_atkey_metadata *metadata) {
+  return 9 // :ivNonce:
+         + metadata->ivnonce.olen;
+}
+
+size_t atclient_atkey_metadata_skeenckeyname_strlen(const atclient_atkey_metadata *metadata) {
+  return 15 // :skeEncKeyName:
+         + metadata->skeenckeyname.olen;
+}
+
+size_t atclient_atkey_metadata_skeencalgo_strlen(const atclient_atkey_metadata *metadata) {
+  return 12 // :skeEncAlgo:
+         + metadata->skeencalgo.olen;
+}
+
+int atclient_atkey_metadata_to_protocol_str(const atclient_atkey_metadata *metadata, char *metadatastr,
+                                            const size_t metadatastrlen, size_t *metadatastrolen) {
+  int ret = 1;
+  size_t pos = 0;
+  size_t len = atclient_atkey_metadata_protocol_strlen(metadata);
+  if (len > metadatastrlen) {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "metadatastr buffer too small: %lu > %lu\n", len,
+                          metadatastrlen);
+    return 1;
+  }
```

Comment:
We use the protocol_strlen again to validate that the provided buffer is large enough.